### PR TITLE
Feature/codegen gather indices greater than rank 1

### PR DESCRIPTION
**Pull request by alteredoxide**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [ ] Made sure the book is up to date with changes in this PR.
    - **NA:** there was nothing in the book regarding gather to begin with

### Related Issues/PRs

- #2023 (where convo began)
- #2192 (most relevant)

### Changes

- Adds support for index tensors of rank > 1 for `Gather` nodes in the burn-import crate.
- The relevant "gather" py models have been modified accordingly, and new test rust models generated.
- Added several tests to `crates/burn-import/src/burn/node/gather.rs` for sanity checks that are faster to than run-checks, but they aren't strictly necessary.

### Testing

- Ran `./run-checks all`
- Ran `cargo test -p burn-import -- burn::node::gather`, which ran the newly added tests for multiple cases:
    - **`test_codegen_gather_idx_1d`:** tests codegen on 1d index input
    - **`test_codegen_gather_idx_2d`:** tests codegen on 2d index input
    - **`test_codegen_gather_shape_input`:** tests codegen for case of a shape input instead of data tensor
    - **`test_codegen_gather_scalar_idx`:** tests codegen for case of the index being passed as scalar instead of tensor
    - **`test_gather_tensor`:** tests actual functionality of gather on input tensor using a 2D index tensor
    - **`test_gather_scalar_idx`:** like `test_gather_tensor` but with scalar index
    - **`test_gather_shape_input`:** tests functionality for shape input instead of data tensor

### CODE REVIEW

File: crates/burn-import/src/burn/node/gather.rs
Comment (laggui) [inline crates/burn-import/src/burn/node/gather.rs:67]: This is not required for `burn-import` crate which doesn't support no_std

File: crates/burn-import/src/burn/node/gather.rs
Comment (laggui) [inline crates/burn-import/src/burn/node/gather.rs:56]: Instead of having two `match` statements on `self.index` one after the other, we should combine the logic for the match arms.

File: crates/burn-import/onnx-tests/tests/gather/gather.py
Comment (laggui) [inline crates/burn-import/onnx-tests/tests/gather/gather.py:1]: I think we should keep the gather 1d test case as before since they have different logic and add a new test for gather nd.

File: crates/burn-import/onnx-tests/tests/onnx_tests.rs
Comment (laggui) [inline crates/burn-import/onnx-tests/tests/onnx_tests.rs:462]: (repeating myself here) let's keep the existing test for gather_1d and add a new test for gather_nd 🙂

File: crates/burn-import/src/burn/node/gather.rs
Comment (laggui) [inline crates/burn-import/src/burn/node/gather.rs:80]: Let's refactor the vector creation to a map + collect instead:

```rust
let out = index_flat.iter_dim(0).map(|idxs| Tensor::select(#input.clone(), #dim, idxs.squeeze::<1>(0))).collect();
```

**Review by laggui — CHANGES_REQUESTED** [review]
Great job! 💪 The ONNX import with codegen is not necessarily the most straightforward part of the codebase to jump into without any prior knowledge 😅 The implementation looks good!

I just have a couple of changes (mostly regarding the tests removed/added, and one to simplify the codegen conditions for the node).

**Comment by laggui** [comment]
Ahh I guess it would depend on the burn config in the project! If you have default features disabled (so, no std) then it might be required.

In this case, you're right. So we should probably use alloc, but instead of having it in the codegen I'm pretty sure we could add it conditionally in the `register_imports` (see for example [`Squeeze`](https://github.com/tracel-ai/burn/blob/main/crates/burn-import/src/burn/node/unsqueeze.rs#L49)).

**Comment by alteredoxide** [comment]
Cool, I'll check that out, thanks. I was wondering where/how imports were set for codegen.

**Comment by alteredoxide** [comment]
@laggui A couple notes:
1. I can confirm that the build fails without using alloc:
```
error[E0433]: failed to resolve: use of undeclared type `Vec`
  --> /home/ao/projects/third-party/rust/burn/target/debug/build/onnx-tests-6ab92c211618b15e/out/model/gather.rs:45:23
   |
45 |         let mut out = Vec::new();
   |                       ^^^ use of undeclared type `Vec`

For more information about this error, try `rustc --explain E0433`.
error: could not compile `onnx-tests` (test "onnx_tests") due to 1 previous error
warning: build failed, waiting for other jobs to finish...
```
2. Registering imports as follows results in ``use of undeclared crate or module `alloc` ``:
```
    fn register_imports(&self, imports: &mut BurnImports) {
        match &self.index {
            Type::Tensor(_) => {
                imports.register("alloc::vec::Vec");
            },
            _ => {}
        }
    }
```

I also tried adding `imports.register("extern crate alloc")`, just in case, but it didn't like that.
Any suggestions?

**Comment by laggui** [comment]
@alteredoxide Actually, if you use the suggested `iter_dim` + `map` + `collect` implementation instead, you don't need to import `Vec`. And the type hint (e.g., `collect::<Vec<_>>()`) shouldn't be required since the compiler will infer the type from the `Tensor::stack` signature.

Give it a shot and let me know how that goes 🙂

**Comment by alteredoxide** [comment]
@laggui Funny, the moment I submitted my last message, I had a feeling the collect was going to come up. I hadn't forgotten about it; I was just testing each change as I made them. But you're right, that solved the issue!

I'm pretty sure I made all requested changes, and tests + run_checks are green. Lmk if you want any further tweaks.

File: crates/burn-import/onnx-tests/build.rs
Comment (laggui) [inline crates/burn-import/onnx-tests/build.rs:35]: If we're not using this file anymore due to the rename, we should remove it from the workspace 🙂

**Review by laggui — APPROVED** [review]
Thanks for applying the changes! 👍 

Don't forget to remove the (now unused) `gather.onnx` file from the workspace now that the 1d and 2d cases have been renamed.

I'm approving in advance 🙂

**Comment by alteredoxide** [comment]
@laggui `gather.onnx` has been deleted and change pushed... I don't have write access, so I can't merge. Thanks!

**Pull request merged**
